plugins/producer_plugin/include/eosio/producer_plugin/production_pause_vote_tracker.hpp

```diff
@@ -0,0 +1,229 @@
+#pragma once
+
+#include <eosio/chain/thread_utils.hpp>
+#include <fc/time.hpp>
+#include <atomic>
+
+namespace eosio {
+
+class production_pause_vote_tracker {
+public:
+   // By default, start with maximum vote_timeout so that production pause mechanism is effectively disabled.
+   // Pre-conditions: vote_timeout and block_acceptance_tolerance should both not be negative.
+   explicit production_pause_vote_tracker(fc::microseconds block_acceptance_tolerance = fc::milliseconds(250),
+                                          fc::microseconds vote_timeout               = fc::microseconds::maximum())
+      : negative_vote_timeout(), block_acceptance_delta(), latest_other_vote(), latest_producer_vote() {
+      set_vote_timeout(vote_timeout);
+      set_block_acceptance_tolerance(block_acceptance_tolerance);
+   }
+
+   // Can be called concurrently with all member functions except:
+   //    + set_vote_timeout
+   //    + check_pause_status
+   //
+   // Sets the vote timeout duration which drives the production pause status.
+   //
+   // Pre-condition: vote_timeout should not be negative.
+   void set_vote_timeout(fc::microseconds vote_timeout) {
+      if (vote_timeout.count() > 0) {
+         negative_vote_timeout = fc::microseconds(-vote_timeout.count());
+      }
+   }
+
+   // Can be called concurrently with all member functions except:
+   //    + set_block_acceptance_tolerance
+   //    + record_received_block
+   //
+   // Sets the block acceptance tolerance duration which determines the relevancy of blocks to record.
+   //
+   // votes can be signaled before accepted_block_header, accepted_block_header is currently only signaled from the
+   // main thread while votes can be signaled off the main thread. This allows votes to be signaled for a block before
+   // the block is signaled. We could track last accepted block id and correlate that with the vote signal, but simpler
+   // to add a signal_tolerance. Normally the tolerance only needs to be a few milliseconds, but no real harm in making
+   // it larger. Half a block interval is a nice value.
+   //
+   // Pre-condition: block_acceptance_tolerance should not be negative.
+   void set_block_acceptance_tolerance(fc::microseconds block_acceptance_tolerance) {
+      assert(block_acceptance_tolerance.count() >= 0);
+
+      if (block_acceptance_tolerance.count() > 0) {
```

Comment:
I missed this earlier but this should be `>=`.

Reply:
Why? `>=` seems correct to me. Why set it to `-0` ?

Reply:
It wouldn't change the value in state if `block_acceptance_tolerance` is 0. 0 is a valid value to pass in to the function.